updates/P7.txt

```diff
@@ -0,0 +1,320 @@
+                           Protocol update
+
+Protocol Version: 7
+Builds on Protocol Version: 6
+
+                               Abstract
+
+This document describes the changes in Concordium protocol version 7 compared to
+protocol version 6.
+
+The protocol change has the following main effects
+- redefines smart contract costs, reducing the cost of smart contract calls;
+- adds support for querying the module reference and contract name of a smart
+  contract instance in smart contracts;
+- changes cooldown behavior so that:
+   * when validators and delegators reduce their stake, the reduction is
+     immediately effective for future stake calculations, and the amount of the
+     reduction is looked for a cooldown period;
+   * validators and delegators can make further changes to their stake while
+     they already have stake in cooldown;
+- disallows shielded (or encrypted) transfer transactions (`TransferToEncrypted`, 
+  `EncryptedAmountTransfer` and `EncryptedAmountTransferWithMemo`).
+
+                              Background
+
+Smart contract execution costs are reduced in protocol version 7 because of
+efficiency improvements in the implementation of smart contract execution.
+The revised cost model is based on an execution model where the interpreter
+stack is compiled to registers.
+
+The new protocol version also imposes a limit on the nesting depth of
+inter-contract calls to 384. This is since each live execution frame can
+require up to 32MB of memory, so limiting the number of active execution
+frames in this way safeguards against excessive memory use.
+
+                                * * *
+
+As of protocol version 7, smart contracts are able to query the module
+reference and contract name of a smart contract instance. This can be used
+to identify the code that is behind a smart contract instance, for example as
+a means of authenticating a smart contract before interacting with it. A
+particular motivation for this is to implement a factory pattern, where one
+smart contract (the factory) is responsible for initializing instances of
+another (the product). On Concordium, one smart contract cannot directly
+create instances of another, but it can be used to invoke an initialization
+method on another. The additional queries allow the factory to ensure that
+the contract instance it is initializing is of the correct type.
+
+                                * * *
+
+Prior to protocol version 7, when a validator or a delegator reduces or
```

Comment:
We typically have a supplemental commentary on the proposal, where it would make sense to do this. The hash of this file is used to reference the protocol update itself, so this document is largely intended to be normative.